src/pybind/mgr/dashboard_v2/components/rgw_client.py

```diff
@@ -0,0 +1,206 @@
+# -*- coding: utf-8 -*-
+from __future__ import absolute_import
+
+import re
+from awsauth import S3Auth
+from ..settings import Settings, Options
+from ..components.rest_client import RestClient, RequestException
+from ..tools import build_url, isset
+from .. import mgr, logger
+
+
+class RgwClient(RestClient):
+    _SYSTEM_USERID = None
+    _ADMIN_PATH = None
+    _host = None
+    _port = None
+    _ssl = None
+    _user_instances = {}
+
+    class NoCredentialsException(Exception):
+        def __init__(self):
+            super(RgwClient.NoCredentialsException,
+                  self).__init__('No RGW credentials found')
+
+    @staticmethod
+    def _determine_rgw_addr():
+        service_map = mgr.get('service_map')
+
+        if not isset(service_map, ['services', 'rgw', 'daemons', 'rgw']):
+            msg = 'No RGW found.'
+            raise LookupError(msg)
+
+        daemon = service_map['services']['rgw']['daemons']['rgw']
+        addr = daemon['addr'].split(':')[0]
+        match = re.search(r'port=(\d+)',
+                          daemon['metadata']['frontend_config#0'])
+        if match:
+            port = int(match.group(1))
+        else:
+            msg = 'Failed to determine RGW port'
+            raise LookupError(msg)
+
+        return addr, port
+
+    @staticmethod
+    def _load_settings():
+        if Settings.RGW_API_SCHEME and Settings.RGW_API_ACCESS_KEY and \
+                Settings.RGW_API_SECRET_KEY:
+            if Options.has_default_value('RGW_API_HOST') and \
+                    Options.has_default_value('RGW_API_PORT'):
+                host, port = RgwClient._determine_rgw_addr()
+            else:
+                host, port = Settings.RGW_API_HOST, Settings.RGW_API_PORT
+        else:
+            raise RgwClient.NoCredentialsException()
+
+        RgwClient._host = host
```

Comment:
Why are these attributes defined in this way vs. in the class constructor? This kind of setting attributes can cause undesired effects because they don't stick with the instance, but rather the whole class which can affect other instances. Debugging why that changes or isn't expected is going to be tricky. For example:

    In [1]: class Foo(object):
       ...:     a = None
       ...:
    
    In [2]: Foo.a = 1
    
    In [3]: f = Foo()
    
    In [4]: f.a
    Out[4]: 1
    
    In [5]: Foo.a = 2
    
    In [6]: f.a
    Out[6]: 2